### PR TITLE
[WIP] ActionsMenu

### PR DESCRIPTION
**Pull request by nktssh**

### CODE REVIEW

File: src/components/ActionsMenu/ActionsMenu.driver.ts [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+import style from './ActionsMenu.st.css';
+
+export interface ActionsMenuDriver extends BaseUniDriver {
+  isMobile(): Promise<boolean>;
+}
+
+export const actionsMenuDriverFactory = (
+  base: UniDriver,
+): ActionsMenuDriver => {
+  const stylableUtil = new StylableUnidriverUtil(style);
```
Comment (giladsegal) [inline src/components/ActionsMenu/ActionsMenu.driver.ts]: We usually avoid using `StylableUnidriverUtil` because it causes problems in e2e tests
You can add a `data-` attribute to your component and verify using this attribute

File: src/components/ActionsMenu/ActionsMenu.st.css [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+    border-width: value(menuBorderWidth);
+    border-color: value(menuBorderColor);
+}
+
+.root:mobile {
+    width: 100%;
+}
+
+.root:alignment(left) {
+    text-align: left;
```
Comment (giladsegal) [inline src/components/ActionsMenu/ActionsMenu.st.css]: This isn't working as expected in storybook I think, maybe you should use `direction: rtl / ltr` or `flex`

File: src/components/ActionsMenu/ActionsMenu.tsx [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+
+export enum Alignment {
+  left = 'left',
+  center = 'center',
+  right = 'right',
+}
+
+export interface ActionsMenuProps {
+  alignment?: Alignment;
+  onBlur?(): void;
```
Comment (giladsegal) [inline src/components/ActionsMenu/ActionsMenu.tsx]: Is it unused because it is a WIP?

File: src/components/ActionsMenu/ActionsMenu.tsx [hunk 4]
```diff
@@ -0,0 +1,44 @@
+import * as React from 'react';
+import styles from './ActionsMenu.st.css';
+import { TPAComponentsConsumer } from '../TPAComponentsConfig';
+import { ActionsMenuItem } from './Item/ActionsMenuItem';
+import { ActionsMenuDivider } from './Divider/Divider';
```
Comment (giladsegal) [inline src/components/ActionsMenu/ActionsMenu.tsx]: You should import from `./Divider`
This way we can change the internal folder structure more easily

File: src/components/ActionsMenu/Divider/Divider.st.css [hunk 5]
```diff
@@ -0,0 +1,15 @@
+:import {
```
Comment (giladsegal) [inline src/components/ActionsMenu/Divider/Divider.st.css]: The library already has a `Divider` component, can you use it instead?

File: src/components/ActionsMenu/Item/ActionsMenuItem.st.css [hunk 6]
```diff
@@ -0,0 +1,73 @@
+:import {
+    -st-from: "../../common/formatters.st";
+    -st-named: color, fallback;
+}
+
+
+:vars {
+    /*Defaults*/
+    DefaultItemFont: 16px Avenir;
```
Comment (giladsegal) [inline src/components/ActionsMenu/Item/ActionsMenuItem.st.css]: You should use wix font stack

File: src/components/ActionsMenu/Item/ActionsMenuItem.st.css [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+.root:hover {
+    background-color: rgba(0,0,0, 0.06);
+}
+
+.root:disabled {
+    opacity: 0.3;
+}
+
+.root:disabled:hover {
+    background-color: inherit;
```
Comment (giladsegal) [inline src/components/ActionsMenu/Item/ActionsMenuItem.st.css]: I think you should also set `cursor: not-allowed`

File: src/components/ActionsMenu/Item/ActionsMenuItem.tsx [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+    } = this.props;
+
+    return (
+      <TPAComponentsConsumer>
+        {({ mobile }) => (
+          <div
+            key={content}
+            {...styles(
+              styles.root,
+              { mobile, disabled, withIcon: !!prefixIcon },
```
Comment (giladsegal) [inline src/components/ActionsMenu/Item/ActionsMenuItem.tsx]: Do you really need `withIcon` state?

File: src/components/ActionsMenu/Item/ActionsMenuItem.tsx [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+      subtitle,
+      disabled,
+      onClick,
+      ...rest
+    } = this.props;
+
+    return (
+      <TPAComponentsConsumer>
+        {({ mobile }) => (
+          <div
```
Comment (giladsegal) [inline src/components/ActionsMenu/Item/ActionsMenuItem.tsx]: Please refer to http://w3c.github.io/aria-practices/#menu for accessibility 
I think you shouldn't use `aria-pressed` as it is used for toggle buttons mostly, you should use `role="menu"`, set `aria-disabled` for disabled items, use a `ul` tag and more

File: src/components/ActionsMenu/Item/ActionsMenuItem.st.css [hunk 10]
```diff
@@ -0,0 +1,73 @@
+:import {
```
Comment (giladsegal) [inline src/components/ActionsMenu/Item/ActionsMenuItem.st.css]: It looks like the styles doesn't completely match zeroheight, can you please notify me once you are done with the styling?

File: src/components/ActionsMenu/ActionsMenu.e2e.ts [hunk 11]
```diff
@@ -0,0 +1,30 @@
+import { browser } from 'protractor';
+import {
+  createStoryUrl,
+  waitForVisibilityOf,
+} from 'wix-ui-test-utils/protractor';
+import { actionsMenuTestkitFactory } from '../../testkit/protractor';
+
+/**
+ * For tests containing interactions.
+ * Can be removed if not used.
```
Comment (jonathanadler) [inline src/components/ActionsMenu/ActionsMenu.e2e.ts]: This file can be removed if there are no special cases to test, that require a real dom.
Just to see that it is rendered in a browser is done in the visual tests.

File: src/components/ActionsMenu/ActionsMenu.tsx (hunk 3 above)
Comment (jonathanadler) [reply src/components/ActionsMenu/ActionsMenu.tsx]: I actually told @nataliia-khitsan to add the onblur callback.
I thought it might be needed for a11y, but i'm not sure anymore.
We can remove it for now.

File: src/components/ActionsMenu/Item/ActionsMenuItem.st.css (hunk 6 above)
Comment (jonathanadler) [reply src/components/ActionsMenu/Item/ActionsMenuItem.st.css]: See `Text.st.css` as an example

File: src/components/ActionsMenu/Item/ActionsMenuItem.st.css (hunk 7 above)
Comment (jonathanadler) [reply src/components/ActionsMenu/Item/ActionsMenuItem.st.css]: @giladsegal i'm not sure changing the cursor is necessary.
I don't see it in the spec.
But we can check with the designers

File: src/components/ActionsMenu/ActionsMenu.driver.ts (hunk 1 above)
Comment (nktssh) [reply src/components/ActionsMenu/ActionsMenu.driver.ts]: Fixed.
It was basically code from generator, maybe we need to change it also?

File: src/components/ActionsMenu/Item/ActionsMenuItem.st.css (hunk 7 above)
Comment (nktssh) [reply src/components/ActionsMenu/Item/ActionsMenuItem.st.css]: @jonathanadler any updates on this?

Comment (jonathanadler) [reply src/components/ActionsMenu/Item/ActionsMenuItem.st.css]: No need to change cursor

File: src/components/ActionsMenu/ActionsMenu.st.css (hunk 2 above)
Comment (jonathanadler) [inline src/components/ActionsMenu/ActionsMenu.st.css]: As @giladsegal wrote, this isn't working as excepted.
However, using `text-align` is fine in this case, since this is the component's alignment, and not the text-direction.
Try to solve why this isn't working correctly.

Comment (jonathanadler) [reply src/components/ActionsMenu/ActionsMenu.st.css]: I think you need to import the `ActionsMenuItem.st.css` file here, and pass the alignment to the component.

**Comment by nktssh** [comment]
@jonathanadler alignment is working now. I've tried different approaches and it seems only this one fulfill the specs (e.g. entire row should be clickable and has custom `:hover` state)

Regarding styling on screenshot attached and in story itself - I've asked previously https://github.com/wix/wix-ui-tpa/issues/103
and got an answer 
> we are aware that this is happening
> it seems to be a race condition inside the story itself that's causing this, and not a bug in the component.
> We will try to find and fix the root cause of this.

Also, I'm not really sure how to test editor overrides (like font and colors) properly - is there an example of it?

Please, point me out if something need to be tuned and in which way! 🙏

File: src/components/ActionsMenu/Item/ActionsMenuItem.tsx (hunk 9 above)
Comment (nktssh) [reply src/components/ActionsMenu/Item/ActionsMenuItem.tsx]: Add mentioned comments addressed, but I'm not sure what's behind "more". 
Please ping me if something else need to be changed

File: src/components/ActionsMenu/ActionsMenu.visual.tsx [hunk 12]
```diff
@@ -0,0 +1,122 @@
...
+            subtitle: 'Subtitle',
+            prefixIcon: <ShareIcon />,
+          }),
+        ],
+      },
+      {
+        it: 'mobile',
+        props: {
+          mobile: true,
+        },
```
Comment (jonathanadler) [inline src/components/ActionsMenu/ActionsMenu.visual.tsx]: We will have to add children to the mobile test, otherwise it will just be empty and the visual test will be redundant.

File: src/components/ActionsMenu/ActionsMenu.st.css [hunk 13]
```diff
@@ -0,0 +1,59 @@
...
+.item {
+    -st-extends: ActionMenuItem;
+}
+
+.root:mobile {
+    width: 100%;
+}
+
+.root:alignment(left) .item {
+    justify-content: flex-start;
```
[the remark this answers is not in the document]
Comment (nktssh) [reply src/components/ActionsMenu/ActionsMenu.st.css]: `Text-align` is fine for text itself but not working with images. 
I've done this by passing alignment down to `Item` and used absolute positioning

[the remark this answers is not in the document]
Comment (nktssh) [reply src/components/ActionsMenu/ActionsMenu.st.css]: I've changed code to avoid passing alignment, so now it looks better 🙂

**Comment by jonathanadler** [comment]
I see that you've fixed the tests.
That's great 👍 
But there are still problems with the styling of the components.
Please go over the design spec (here: https://zeroheight.com/7sjjzhgo2/p/05a744/b/46537f/t/53af3f) and once it's done send me a request for review

File: src/components/ActionsMenu/ActionsMenu.tsx [hunk 14]
```diff
@@ -0,0 +1,60 @@
...
+    alignment: Alignment.left,
+  };
+
+  render() {
+    const { alignment, children, ...rest } = this.props;
+    return (
+      <TPAComponentsConsumer>
+        {({ mobile }) => (
+          <ul
+            key={alignment}
```
Comment (jonathanadler) [inline src/components/ActionsMenu/ActionsMenu.tsx]: I think the `key` here isn't necessary.

File: src/components/ActionsMenu/ActionsMenu.tsx [hunk 15]
```diff
@@ -0,0 +1,60 @@
...
+
+export interface ActionsMenuProps {
+  alignment?: Alignment;
+}
+
+interface DefaultProps {
+  alignment: Alignment;
+}
+
+function getDataAttributes(mobile: boolean) {
```
Comment (jonathanadler) [inline src/components/ActionsMenu/ActionsMenu.tsx]: I think it's better to have this function as a class method, no?

File: src/components/ActionsMenu/ActionsMenu.st.css [hunk 16]
```diff
@@ -0,0 +1,65 @@
...
+    padding: 12px 0;
+    max-width: 320px;
+    background-color: value(menuBackgroundColor);
+    border-style: solid;
+    border-width: value(menuBorderWidth);
+    border-color: value(menuBorderColor);
+}
+
+.item {
+    -st-extends: ActionMenuItem;
```
Comment (jonathanadler) [inline src/components/ActionsMenu/ActionsMenu.st.css]: I don't think this is needed if we're not overriding anything here.

File: src/components/ActionsMenu/ActionsMenu.tsx [hunk 17]
```diff
@@ -0,0 +1,60 @@
...
+            key={alignment}
+            {...styles('root', { mobile, alignment }, rest)}
+            {...getDataAttributes(mobile)}
+            role="menu"
+            tabIndex={-1}
+          >
+            {React.Children.map(children, (child: React.ReactElement) =>
+              child.type === ActionsMenuItem
+                ? React.cloneElement(child, { className: styles.item })
+                : React.cloneElement(child, { className: styles.divider }),
```
Comment (jonathanadler) [inline src/components/ActionsMenu/ActionsMenu.tsx]: I actually think that passing the classNames aren't really necessary here.
I think the static Divider can be a component that wraps the `wix-ui-tpa/Divider`, and adds the necessary padding.

File: src/components/ActionsMenu/ActionsMenu.visual.tsx [hunk 18]
```diff
@@ -0,0 +1,122 @@
...
+            content: 'item 1',
+          }),
+          generateItem({
+            content: 'item 3',
+            subtitle: 'Subtitle 2',
+          }),
+        ],
+      },
+    ],
+  },
```
Comment (jonathanadler) [inline src/components/ActionsMenu/ActionsMenu.visual.tsx]: Let's add tests for all the different alignments: `left`, `center` & `right`.

File: src/components/ActionsMenu/Item/ActionsMenuItem.st.css [hunk 19]
```diff
@@ -0,0 +1,77 @@
...
+    -st-states: disabled, mobile;
+    height: 40px;
+    padding: 8px 24px;
+    display: flex;
+    cursor: pointer;
+    position: relative;
+}
+
+.root:hover {
+    background-color: rgba(0,0,0, 0.06);
```
Comment (jonathanadler) [inline src/components/ActionsMenu/Item/ActionsMenuItem.st.css]: This should be used with `color-5`:
```
applyOpacity(color(color-5), 0.06);
```

File: src/components/ActionsMenu/Item/ActionsMenuItem.st.css [hunk 20]
```diff
@@ -0,0 +1,77 @@
...
+.root:disabled:hover {
+    background-color: inherit;
+}
+
+.root .text {
+    display: flex;
+    flex-direction: column;
+    color: color(color-5);
+    width: 100%;
+}
```
Comment (jonathanadler) [inline src/components/ActionsMenu/Item/ActionsMenuItem.st.css]: Theoretically, a more "correct stylable" way to write this is by extending the `Text` component:
```css
:import {
  -st-from: '../../Text/Text.st.css';
  -st-default: Text;
}

.text {
  -st-extends: Text;
  -st-mixin: Text(
    MainTextColor color(color-5)
  );
  ...
}
```
(by the way, the `.root` isn't needed here, stylable knows to deal with it)

File: src/components/ActionsMenu/Item/ActionsMenuItem.st.css [hunk 21]
```diff
@@ -0,0 +1,77 @@
...
+
+.root:mobile .content {
+    font: fallback(font(value(ItemFont)), font(value(ItemFontMobile)));
+}
+
+.root .subtitle {
+    font: fallback(font(value(ItemSubtitleFont)), font(value(ItemSubtitleFontDesktop)));
+    color: applyOpacity(color(color-5), 0.6);
+    width: 100%;
```
Comment (jonathanadler) [inline src/components/ActionsMenu/Item/ActionsMenuItem.st.css]: Same as I've written above 👆 

```css
.subtitle {
  -st-extends: Text;
  -st-mixin: Text(
    MainTextColor applyOpacity(color(color-5), 0.6),
    MainTextFont fallback(font(value(ItemSubtitleFont)), font(value(ItemSubtitleFontDesktop)))
  );
  ...
}
```

File: src/components/ActionsMenu/Item/ActionsMenuItem.tsx [hunk 22]
```diff
@@ -0,0 +1,58 @@
...
+    return (
+      <TPAComponentsConsumer>
+        {({ mobile }) => (
+          <li
+            key={content}
+            {...styles('root', { mobile, disabled }, rest)}
+            role="menuitem"
+            tabIndex={-1}
+            aria-disabled={disabled}
+            onClick={disabled ? noop : onClick}
```
Comment (jonathanadler) [inline src/components/ActionsMenu/Item/ActionsMenuItem.tsx]: No need to use `noop`.
It's better to pass `undefined` so that no function will be attached to the event.

File: src/components/ActionsMenu/README.md [hunk 23]
```diff
@@ -0,0 +1,25 @@
...
+Child of `ActionsMenu` must be either `Divider` or `ActionsMenuItem`, see Examples section for details.
+
+
+The `ActionsMenu` implementation provides a few default overridable styles. By default site theme is used.
+
+## Theme properties (all are optional)
+| propName              | propType | description                                                |
+|-----------------------|----------|------------------------------------------------------------|
+| MenuBackgroundColor   | string   | The color of the background instead of the default color   |
+| MenuBorderWidth       | string   | The color of the menu width instead of the default one     |
```
Comment (jonathanadler) [inline src/components/ActionsMenu/README.md]: It should be:
`The width of the border...`

File: src/components/ActionsMenu/dataHooks.ts [hunk 24]
```diff
@@ -0,0 +1,3 @@
+export enum ACTIONS_MENU__DATA_KEYS {
```
Comment (jonathanadler) [inline src/components/ActionsMenu/dataHooks.ts]: A bit picky, but we don't use `__` in our other `dataHooks` files.
So you can use just one `_`, for consistency.

File: src/components/ActionsMenu/ActionsMenu.st.css [hunk 25]
```diff
@@ -0,0 +1,65 @@
...
+    margin: 10px 13px;
+    width: auto;
+}
+
+.root:mobile {
+    width: 100%;
+}
+
+.root:alignment(left) .item::container {
+    left: 24px;
```
[the remark this answers is not in the document]
Comment (nktssh) [reply src/components/ActionsMenu/ActionsMenu.st.css]: to imitate the padding, since we absolutely position here

File: src/components/ActionsMenu/ActionsMenu.tsx (hunk 15 above)
Comment (nktssh) [reply src/components/ActionsMenu/ActionsMenu.tsx]: now it has no deps from props or class instance, so I've put it aside.
Could do it as static method though

File: src/components/ActionsMenu/ActionsMenu.st.css (hunk 16 above)
Comment (nktssh) [reply src/components/ActionsMenu/ActionsMenu.st.css]: without it alignment would not work

File: src/components/ActionsMenu/Item/ActionsMenuItem.st.css (hunk 20 above)
Comment (nktssh) [reply src/components/ActionsMenu/Item/ActionsMenuItem.st.css]: ok, will use `Text`
but in that case `.root` is required because we get an error otherwise

File: src/components/ActionsMenu/Item/ActionsMenuItem.st.css (hunk 21 above)
Comment (nktssh) [reply src/components/ActionsMenu/Item/ActionsMenuItem.st.css]: fixed

File: src/components/ActionsMenu/README.md (hunk 23 above)
Comment (nktssh) [reply src/components/ActionsMenu/README.md]: fixed

File: src/components/ActionsMenu/ActionsMenu.tsx (hunk 15 above)
Comment (nktssh) [reply src/components/ActionsMenu/ActionsMenu.tsx]: moved to class

File: src/components/ActionsMenu/ActionsMenu.tsx (hunk 14 above)
Comment (nktssh) [reply src/components/ActionsMenu/ActionsMenu.tsx]: removed

File: src/components/ActionsMenu/Item/ActionsMenuItem.st.css (hunk 20 above)
Comment (jonathanadler) [reply src/components/ActionsMenu/Item/ActionsMenuItem.st.css]: That's weird 🤔

File: src/components/ActionsMenu/Item/ActionsMenuItem.st.css (hunk 19 above)
Comment (jonathanadler) [reply src/components/ActionsMenu/Item/ActionsMenuItem.st.css]: This still needs to change

File: src/components/ActionsMenu/ActionsMenu.tsx (hunk 17 above)
Comment (jonathanadler) [reply src/components/ActionsMenu/ActionsMenu.tsx]: What about this?

Comment (nktssh) [reply src/components/ActionsMenu/ActionsMenu.tsx]: resolved

File: src/components/ActionsMenu/Item/ActionsMenuItem.st.css (hunk 19 above)
Comment (nktssh) [reply src/components/ActionsMenu/Item/ActionsMenuItem.st.css]: fixed

**Comment by nktssh** [comment]
@jonathanadler all the comments were fixed or commented

File: src/components/ActionsMenu/Item/ActionsMenuItem.st.css [hunk 26]
```diff
@@ -0,0 +1,97 @@
...
+}
+
+.root .text {
+    -st-extends: Text;
+    -st-mixin: Text(
+        MainTextColor color(color-5)
+    );
+    display: flex;
+    flex-direction: column;
+    width: 100%;
```
Comment (jonathanadler) [inline src/components/ActionsMenu/Item/ActionsMenuItem.st.css]: This should also have a `justify-content: center;` since we want the text to always be in the middle of the item

File: src/components/ActionsMenu/Item/ActionsMenuItem.st.css [hunk 27]
```diff
@@ -0,0 +1,97 @@
...
+}
+
+.root {
+    -st-states: disabled, mobile;
+    height: 40px;
+    padding: 8px 24px;
+    display: flex;
+    cursor: pointer;
+    position: relative;
+}
```
Comment (jonathanadler) [inline src/components/ActionsMenu/Item/ActionsMenuItem.st.css]: Take a look at this snapshot I took:
image
Note that you can see the main container's border.
And we need the hover background to take all the width (including the border).
So we need to have the item take `width: calc(100% + value(parentsBorder));`
And for that to work we also need to add `box-sizing: border-box` to the ActionsMenu .root
